pkg/storegateway/bucket_index_reader.go

```diff
@@ -628,43 +631,44 @@ func (r *bucketIndexReader) preloadSeries(ctx context.Context, ids []storage.Ser
 func (r *bucketIndexReader) loadSeries(ctx context.Context, ids []storage.SeriesRef, refetch bool, start, end uint64, loaded *bucketIndexLoadedSeries, stats *safeQueryStats) error {
 	begin := time.Now()
 
-	b, err := r.block.readIndexRange(ctx, int64(start), int64(end-start))
+	reader, err := r.block.indexRangeReader(ctx, int64(start), int64(end-start))
 	if err != nil {
 		return errors.Wrap(err, "read series range")
 	}
-
-	stats.update(func(stats *queryStats) {
-		stats.seriesFetchCount++
-		stats.seriesFetched += len(ids)
-		stats.seriesFetchDurationSum += time.Since(begin)
-		stats.seriesFetchedSizeSum += int(end - start)
-	})
+	defer runutil.CloseWithLogOnErr(r.block.logger, reader, "loadSeries close range reader")
+	byteReader := &uvarintSequenceReader{r: reader, offset: start}
 
 	for i, id := range ids {
-		c := b[uint64(id)-start:]
-
-		l, n := binary.Uvarint(c)
-		if n < 1 {
-			return errors.New("reading series length failed")
+		size, err := byteReader.Uvarint(uint64(id))
+		if err != nil {
+			return err
 		}
-		if len(c) < n+int(l) {
+		seriesBytes := loaded.bytesPool.Get(int(size))
+		n, err := io.ReadFull(byteReader, seriesBytes)
+		if n != int(size) || (err != nil && errors.Is(err, io.ErrUnexpectedEOF)) {
 			if i == 0 && refetch {
-				return errors.Errorf("invalid remaining size, even after refetch, remaining: %d, expected %d", len(c), n+int(l))
+				return errors.Errorf("invalid remaining size, even after refetch, read %d, expected %d", n, size)
 			}
 
 			// Inefficient, but should be rare.
 			r.block.metrics.seriesRefetches.Inc()
-			level.Warn(r.block.logger).Log("msg", "series size exceeded expected size; refetching", "id", id, "series length", n+int(l), "maxSeriesSize", maxSeriesSize)
-
+			level.Warn(r.block.logger).Log("msg", "series size exceeded expected size; refetching", "series_id", id, "series_length", size, "max_series_size", maxSeriesSize)
 			// Fetch plus to get the size of next one if exists.
-			return r.loadSeries(ctx, ids[i:], true, uint64(id), uint64(id)+uint64(n+int(l)+1), loaded, stats)
+			return r.loadSeries(ctx, ids[i:], true, uint64(id), uint64(id)+binary.MaxVarintLen64+size+1, loaded, stats)
+		} else if err != nil {
+			return errors.Wrapf(err, "fetching series %d from block %s", id, r.block.meta.ULID)
 		}
-		c = c[n : n+int(l)]
+		loaded.addSeries(id, seriesBytes)
 
-		loaded.addSeries(id, c)
-
-		r.block.indexCache.StoreSeriesForRef(r.block.userID, r.block.meta.ULID, id, c)
+		r.block.indexCache.StoreSeriesForRef(r.block.userID, r.block.meta.ULID, id, seriesBytes)
```

Comment:
`seriesBytes` is loaded from a pool, but `StoreSeriesForRef()` is async and there's no guarantee that when it gets released to the pool the cache also has done with that.

Reply:
oof, nice catch. I removed the pooling. Benchmarks show increased allocated objects (since we do a byte slice per series now).

<details>
<summary>pooled vs unpooled series</summary>

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/mimir/pkg/storegateway
                                                                                                   │ after-8e7626872.txt │         after-c1b1d4f7a.txt         │
                                                                                                   │       sec/op        │    sec/op     vs base               │
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X"-10                                           12.04µ ± ∞ ¹   12.29µ ± ∞ ¹   +2.06% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j="foo"-10                                   23.12µ ± ∞ ¹   23.75µ ± ∞ ¹   +2.76% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j!="foo"-10                                  23.08µ ± ∞ ¹   25.43µ ± ∞ ¹  +10.17% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"XXX|YYY"-10                                    1.859µ ± ∞ ¹   1.867µ ± ∞ ¹        ~ (p=0.143 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"X.+"-10                                        1.829µ ± ∞ ¹   1.828µ ± ∞ ¹        ~ (p=1.000 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"X|Y|Z"-10                                      31.47µ ± ∞ ¹   31.65µ ± ∞ ¹   +0.58% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1"-10                                            1.306 ± ∞ ¹    1.234 ± ∞ ¹   -5.52% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j="foo"-10                                    1.218 ± ∞ ¹    1.214 ± ∞ ¹        ~ (p=0.886 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n="1"-10                                    1.230 ± ∞ ¹    1.213 ± ∞ ¹        ~ (p=0.486 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j!="foo"-10                                   1.222 ± ∞ ¹    1.202 ± ∞ ¹        ~ (p=0.114 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*"-10                                          3.913 ± ∞ ¹    3.852 ± ∞ ¹        ~ (p=0.686 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".+"-10                                          5.244 ± ∞ ¹    5.168 ± ∞ ¹        ~ (p=0.886 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"^.+$",j=~"X.+"-10                              8.175m ± ∞ ¹   7.840m ± ∞ ¹   -4.09% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~""-10                                            1.517 ± ∞ ¹    1.482 ± ∞ ¹   -2.31% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!=""-10                                            5.254 ± ∞ ¹    5.177 ± ∞ ¹        ~ (p=0.114 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",j="foo"-10                            1.228 ± ∞ ¹    1.216 ± ∞ ¹        ~ (p=0.114 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",j="foo"-10                         23.80µ ± ∞ ¹   23.36µ ± ∞ ¹   -1.86% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",i!="2",j="foo"-10                     1.239 ± ∞ ¹    1.225 ± ∞ ¹   -1.12% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!=""-10                                      2.559 ± ∞ ¹    2.475 ± ∞ ¹   -3.27% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j="foo"-10                              2.596 ± ∞ ¹    2.447 ± ∞ ¹   -5.74% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"X.+"-10                            7.168m ± ∞ ¹   7.081m ± ∞ ¹        ~ (p=0.057 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"XXX|YYY"-10                        15.66µ ± ∞ ¹   15.26µ ± ∞ ¹   -2.54% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                        55.51µ ± ∞ ¹   54.90µ ± ∞ ¹        ~ (p=0.200 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",j="foo"-10                            2.535 ± ∞ ¹    2.454 ± ∞ ¹   -3.17% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"1.+",j="foo"-10                          291.7m ± ∞ ¹   278.7m ± ∞ ¹   -4.44% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!="2",j="foo"-10                     2.511 ± ∞ ¹    2.460 ± ∞ ¹   -2.03% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10                   2.597 ± ∞ ¹    2.461 ± ∞ ¹   -5.22% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10            23.73µ ± ∞ ¹   23.48µ ± ∞ ¹        ~ (p=0.057 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"0xxx|1xxx|2xxx"-10                             255.3µ ± ∞ ¹   260.6µ ± ∞ ¹   +2.08% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"(0|1|2)xxx"-10                                 270.7µ ± ∞ ¹   258.7µ ± ∞ ¹        ~ (p=0.057 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"[0-2]xxx"-10                                   257.0µ ± ∞ ¹   258.8µ ± ∞ ¹        ~ (p=0.686 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!~[0-2]xxx-10                                      4.009 ± ∞ ¹    3.878 ± ∞ ¹        ~ (p=0.200 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*",_i!~[0-2]xxx-10                             4.076 ± ∞ ¹    3.944 ± ∞ ¹        ~ (p=0.343 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"<unique_prefix>.+"-10                          374.5µ ± ∞ ¹   345.0µ ± ∞ ¹        ~ (p=0.114 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"<unique_prefix>.+"-10                    1.585m ± ∞ ¹   1.516m ± ∞ ¹   -4.33% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!~"<unique_prefix>.+"-10                     1.267 ± ∞ ¹    1.235 ± ∞ ¹   -2.46% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10          11.95m ± ∞ ¹   11.86m ± ∞ ¹        ~ (p=0.886 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10          41.14m ± ∞ ¹   40.72m ± ∞ ¹   -1.01% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10          11.47m ± ∞ ¹   11.45m ± ∞ ¹        ~ (p=0.686 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10          390.0µ ± ∞ ¹   364.0µ ± ∞ ¹   -6.69% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10          59.52m ± ∞ ¹   56.40m ± ∞ ¹   -5.26% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/p!=""-10                                            1.714 ± ∞ ¹    1.652 ± ∞ ¹   -3.66% (p=0.029 n=4)
geomean                                                                                                     17.56m         17.22m         -1.90%
¹ need >= 6 samples for confidence interval at level 0.95

                                                                                                   │ after-8e7626872.txt │          after-c1b1d4f7a.txt          │
                                                                                                   │        B/op         │     B/op       vs base                │
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X"-10                                          2.230Ki ± ∞ ¹   2.230Ki ± ∞ ¹       ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j="foo"-10                                  2.934Ki ± ∞ ¹   2.935Ki ± ∞ ¹       ~ (p=0.629 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j!="foo"-10                                 2.934Ki ± ∞ ¹   2.934Ki ± ∞ ¹       ~ (p=1.000 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"XXX|YYY"-10                                   1.852Ki ± ∞ ¹   1.852Ki ± ∞ ¹       ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"X.+"-10                                       1.797Ki ± ∞ ¹   1.797Ki ± ∞ ¹       ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"X|Y|Z"-10                                     2.918Ki ± ∞ ¹   2.918Ki ± ∞ ¹       ~ (p=1.000 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1"-10                                          168.3Mi ± ∞ ¹   174.2Mi ± ∞ ¹  +3.50% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j="foo"-10                                  116.5Mi ± ∞ ¹   119.2Mi ± ∞ ¹  +2.35% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n="1"-10                                  116.6Mi ± ∞ ¹   119.3Mi ± ∞ ¹  +2.28% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j!="foo"-10                                 116.8Mi ± ∞ ¹   119.3Mi ± ∞ ¹  +2.14% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*"-10                                        3.076Gi ± ∞ ¹   3.226Gi ± ∞ ¹  +4.86% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".+"-10                                        3.230Gi ± ∞ ¹   3.378Gi ± ∞ ¹  +4.60% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"^.+$",j=~"X.+"-10                             10.69Mi ± ∞ ¹   10.69Mi ± ∞ ¹       ~ (p=0.114 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~""-10                                          229.3Mi ± ∞ ¹   229.3Mi ± ∞ ¹       ~ (p=0.486 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!=""-10                                          3.230Gi ± ∞ ¹   3.379Gi ± ∞ ¹  +4.61% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",j="foo"-10                          119.7Mi ± ∞ ¹   122.4Mi ± ∞ ¹  +2.29% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",j="foo"-10                        3.106Ki ± ∞ ¹   3.106Ki ± ∞ ¹       ~ (p=1.000 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",i!="2",j="foo"-10                   121.2Mi ± ∞ ¹   123.7Mi ± ∞ ¹  +2.07% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!=""-10                                    359.6Mi ± ∞ ¹   365.4Mi ± ∞ ¹  +1.60% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j="foo"-10                            308.1Mi ± ∞ ¹   310.9Mi ± ∞ ¹  +0.89% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"X.+"-10                           10.69Mi ± ∞ ¹   10.69Mi ± ∞ ¹       ~ (p=0.486 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"XXX|YYY"-10                       3.826Ki ± ∞ ¹   3.826Ki ± ∞ ¹       ~ (p=1.000 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                       5.266Ki ± ∞ ¹   5.266Ki ± ∞ ¹       ~ (p=1.000 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",j="foo"-10                          308.3Mi ± ∞ ¹   310.9Mi ± ∞ ¹  +0.83% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"1.+",j="foo"-10                         48.93Mi ± ∞ ¹   49.11Mi ± ∞ ¹  +0.38% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!="2",j="foo"-10                   309.3Mi ± ∞ ¹   312.2Mi ± ∞ ¹  +0.93% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10                 315.4Mi ± ∞ ¹   317.7Mi ± ∞ ¹  +0.73% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10           3.310Ki ± ∞ ¹   3.310Ki ± ∞ ¹       ~ (p=0.429 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"0xxx|1xxx|2xxx"-10                            105.3Ki ± ∞ ¹   110.1Ki ± ∞ ¹  +4.50% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"(0|1|2)xxx"-10                                105.2Ki ± ∞ ¹   109.7Ki ± ∞ ¹  +4.27% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"[0-2]xxx"-10                                  105.3Ki ± ∞ ¹   109.8Ki ± ∞ ¹  +4.33% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!~[0-2]xxx-10                                    3.073Gi ± ∞ ¹   3.222Gi ± ∞ ¹  +4.84% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*",_i!~[0-2]xxx-10                           3.076Gi ± ∞ ¹   3.225Gi ± ∞ ¹  +4.84% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"<unique_prefix>.+"-10                         1.747Mi ± ∞ ¹   1.743Mi ± ∞ ¹  -0.21% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"<unique_prefix>.+"-10                   2.961Mi ± ∞ ¹   2.959Mi ± ∞ ¹  -0.09% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!~"<unique_prefix>.+"-10                   170.4Mi ± ∞ ¹   175.9Mi ± ∞ ¹  +3.27% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10         20.14Mi ± ∞ ¹   20.13Mi ± ∞ ¹  -0.05% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10         59.18Mi ± ∞ ¹   59.16Mi ± ∞ ¹  -0.02% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10         14.01Mi ± ∞ ¹   14.00Mi ± ∞ ¹       ~ (p=0.114 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10         1.753Mi ± ∞ ¹   1.750Mi ± ∞ ¹  -0.16% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10         77.53Mi ± ∞ ¹   77.51Mi ± ∞ ¹  -0.02% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/p!=""-10                                          672.9Mi ± ∞ ¹   702.7Mi ± ∞ ¹  +4.43% (p=0.029 n=4)
geomean                                                                                                    5.908Mi         5.997Mi        +1.51%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal

                                                                                                   │ after-8e7626872.txt │          after-c1b1d4f7a.txt          │
                                                                                                   │      allocs/op      │  allocs/op    vs base                 │
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X"-10                                            36.00 ± ∞ ¹    36.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j="foo"-10                                    44.00 ± ∞ ¹    44.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",j!="foo"-10                                   44.00 ± ∞ ¹    44.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"XXX|YYY"-10                                     31.00 ± ∞ ¹    31.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~"X.+"-10                                         31.00 ± ∞ ¹    31.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"X|Y|Z"-10                                       46.00 ± ∞ ¹    46.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1"-10                                           1.410M ± ∞ ¹   1.610M ± ∞ ¹  +14.16% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j="foo"-10                                   1.004M ± ∞ ¹   1.104M ± ∞ ¹   +9.94% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n="1"-10                                   1.004M ± ∞ ¹   1.104M ± ∞ ¹   +9.95% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",j!="foo"-10                                  1.005M ± ∞ ¹   1.104M ± ∞ ¹   +9.94% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*"-10                                         20.88M ± ∞ ¹   25.87M ± ∞ ¹  +23.92% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".+"-10                                         23.63M ± ∞ ¹   28.62M ± ∞ ¹  +21.13% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"^.+$",j=~"X.+"-10                              100.0k ± ∞ ¹   100.0k ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~""-10                                           2.750M ± ∞ ¹   2.750M ± ∞ ¹        ~ (p=0.829 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!=""-10                                           23.63M ± ∞ ¹   28.62M ± ∞ ¹  +21.13% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",j="foo"-10                           1.004M ± ∞ ¹   1.104M ± ∞ ¹   +9.95% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",j="foo"-10                          44.00 ± ∞ ¹    44.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".*",i!="2",j="foo"-10                    1.004M ± ∞ ¹   1.104M ± ∞ ¹   +9.94% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!=""-10                                     4.160M ± ∞ ¹   4.360M ± ∞ ¹   +4.80% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j="foo"-10                             3.754M ± ∞ ¹   3.854M ± ∞ ¹   +2.66% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"X.+"-10                            100.1k ± ∞ ¹   100.1k ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!="",j=~"XXX|YYY"-10                         61.00 ± ∞ ¹    61.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"X|Y|Z",j="foo"-10                         82.00 ± ∞ ¹    82.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",j="foo"-10                           3.754M ± ∞ ¹   3.854M ± ∞ ¹   +2.66% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"1.+",j="foo"-10                          417.4k ± ∞ ¹   428.4k ± ∞ ¹   +2.65% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!="2",j="foo"-10                    3.754M ± ∞ ¹   3.854M ± ∞ ¹   +2.66% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~".+",i!~"2.*",j="foo"-10                  3.948M ± ∞ ¹   4.037M ± ∞ ¹   +2.25% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="X",i=~"^.+$",i!~"^.*2.*$",j="foo"-10             44.00 ± ∞ ¹    44.00 ± ∞ ¹        ~ (p=1.000 n=4) ²
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"0xxx|1xxx|2xxx"-10                              896.0 ± ∞ ¹   1042.0 ± ∞ ¹  +16.29% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"(0|1|2)xxx"-10                                  895.0 ± ∞ ¹   1041.0 ± ∞ ¹  +16.31% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"[0-2]xxx"-10                                    895.0 ± ∞ ¹   1041.0 ± ∞ ¹  +16.31% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i!~[0-2]xxx-10                                     20.88M ± ∞ ¹   25.87M ± ∞ ¹  +23.92% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~".*",_i!~[0-2]xxx-10                            20.88M ± ∞ ¹   25.87M ± ∞ ¹  +23.92% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/i=~"<unique_prefix>.+"-10                           380.0 ± ∞ ¹    427.0 ± ∞ ¹  +12.37% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i=~"<unique_prefix>.+"-10                     266.0 ± ∞ ¹    265.0 ± ∞ ¹   -0.38% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/n="1",i!~"<unique_prefix>.+"-10                    1.410M ± ∞ ¹   1.610M ± ∞ ¹  +14.16% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",p="foo",i=~"<unique_prefix>.+"-10           313.0 ± ∞ ¹    318.5 ± ∞ ¹   +1.76% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".+",i=~"<unique_prefix>.+"-10          1.399k ± ∞ ¹   1.414k ± ∞ ¹   +1.07% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j="foo",n=~".*",i=~"<unique_prefix>.+"-10           328.0 ± ∞ ¹    343.5 ± ∞ ¹   +4.73% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".*",n=~".*",i=~"<unique_prefix>.+"-10           414.0 ± ∞ ¹    461.0 ± ∞ ¹  +11.35% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/j=~".+",n=~".+",i=~"<unique_prefix>.+"-10          1.536k ± ∞ ¹   1.579k ± ∞ ¹   +2.83% (p=0.029 n=4)
OpenBlockSeriesChunkRefsSetsIterator/without_index_cache/p!=""-10                                           4.651M ± ∞ ¹   5.650M ± ∞ ¹  +21.47% (p=0.029 n=4)
geomean                                                                                                     25.36k         27.18k         +7.17%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal
```

</details>

overall results in the PR are still positive - see https://github.com/grafana/mimir/pull/4926#issuecomment-1535234587

I am now trying to reduce allocated objects further by using and _not_ releasing a slab pool